boards: add common configuration for CDC ACM UART

jfischer-no opened the pull request:
Many boards have similar code to configure the USB and CDC ACM UART that
they want to use as a logging or shell backend. Some of them have an
incorrect or incomplete configuration.

These boards do not have a built-in debug adapter, but a SoC with a USB
device controller and a bootloader with USB device support. Introduce
common CDC ACM UART configuration that these boards should use for
logging or shell backend to avoid duplicate or incorrect configuration.

Derived from #81220 and #81228, as first step (or alternative).

jfischer-no committed:
boards: use board common CDC ACM UART configuration

Remove all USB and CDC ACM configuration in favor of common
configuraiton.

Do not adapt board-specific configurations such as unknown PID/VID or
string descriptors. There is no justification for using them on specific
boards, and we do not have formal approval to use them in the project
tree. Also, we need more uniform configuration, since the main reason
for enabling CDC ACM here is to allow users to run examples like
hello_world right out of the box. Of course, anyone is free to customize
these settings in their fork or downstream project.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: add common configuration for CDC ACM UART

Many boards have similar code to configure the USB and CDC ACM UART that
they want to use as a logging or shell backend. Some of them have an
incorrect or incomplete configuration.

These boards do not have a built-in debug adapter, but a SoC with a USB
device controller and a bootloader with USB device support. Introduce
common CDC ACM UART configuration that these boards should use for
logging or shell backend to avoid duplicate or incorrect configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

rerickson1 reviewed (dismissed):
Ezurio boards look ok

nordicjm commented on boards/common/cdc_acm_serial.dts:
```diff
@@ -0,0 +1,21 @@
+};
+
+zephyr_udc0: &usbd {
+	cdc_acm_uart: cdc_acm_uart {
```
should this not be `cdc_acm_uart0` if it supports multiple instances?

jfischer-no replied:
`cdc_acm_uart0` is likely to conflict with samples or user applications, we should rather name it `board_cdc_acm_uart`, like we prefix snippets and shields.

nordicjm replied:
`board_*` prefix sounds good

jfischer-no replied:
I also moved them to `boards/common/usb/`

erwango commented on boards/arduino/opta/Kconfig.defconfig:
```diff
@@ -11,34 +11,6 @@ config NET_L2_ETHERNET
-endif # LOG
-
-endif # USB_DEVICE_STACK
+source "${ZEPHYR_BASE}/boards/common/Kconfig.cdc_acm_serial"
```
Nitpick:
Harmonize path between Kconfig and dts:
`source "../boards/common/Kconfig.cdc_acm_serial"`
`#include <../boards/common/cdc_acm_serial.dts>`
or
`source "${ZEPHYR_BASE}/boards/common/Kconfig.cdc_acm_serial"`
`#include <${ZEPHYR_BASE}/boards/common/cdc_acm_serial.dts>`

jfischer-no replied:
${ZEPHYR_BASE} cannot be resolved in dt.

erwango replied:
Can we then use `../boards/common` for Kconfig ?

erwango commented on boards/st/sensortile_box_pro/Kconfig.defconfig:
```diff
@@ -20,46 +20,7 @@ config SPI_STM32_INTERRUPT
-endif # BOARD_SERIAL_BACKEND_CDC_ACM
+source "${ZEPHYR_BASE}/boards/common/Kconfig.cdc_acm_serial"
 
 DT_CHOSEN_ZEPHYR_CONSOLE := zephyr,console
```
Shouldn't this be cleaned up as well ?

erwango replied:
This and following lines to be cleaned up as now conflicts with configuration provided in `${ZEPHYR_BASE}/boards/common/usb/Kconfig.cdc_acm_serial`

erwango commented on boards/st/sensortile_box_pro/Kconfig.defconfig:
```diff
@@ -20,46 +20,7 @@ config SPI_STM32_INTERRUPT
-config USB_DEVICE_REMOTE_WAKEUP
-	default n
-
-config USB_DEVICE_VID
-	default 0x0483
-
-config USB_DEVICE_PID
-	default 0x5740
-
-config USB_DEVICE_PRODUCT
-	default "Zephyr CDC SensorTile.box PRO"
```
These 3 symbols look like legit board specific configuration.
If there a specific, documented, reason to remove them?

jfischer-no replied:
rather other way around. Added to commit message:
```
    Do not adapt board-specific configurations such as unknown PID/VID or
    string descriptors. There is no justification for using them on specific
    boards, and we do not have formal approval to use them in the project
    tree. Also, we need more uniform configuration, since the main reason
    for enabling CDC ACM here is to allow users to run examples like
    hello_world right out of the box. Of course, anyone is free to customize
    these settings in their fork or downstream project.
```

erwango commented on boards/st/steval_stwinbx1/Kconfig.defconfig:
```diff
@@ -20,46 +20,7 @@ config SPI_STM32_INTERRUPT
-endif # BOARD_SERIAL_BACKEND_CDC_ACM
+source "${ZEPHYR_BASE}/boards/common/usb/Kconfig.cdc_acm_serial"
 
 DT_CHOSEN_ZEPHYR_CONSOLE := zephyr,console
```
Ditto

jfischer-no replied:
removed

jacobw commented on boards/adafruit/feather_nrf52840/Kconfig.defconfig:
```diff
@@ -9,4 +9,6 @@ if BOARD_ADAFRUIT_FEATHER_NRF52840
 config BT_CTLR
 	default BT
 
+source "${ZEPHYR_BASE}/boards/common/usb/Kconfig.cdc_acm_serial"
```
This would enable CDC ACM on the 2 non UF2 variants of this board which currently don't have it enabled. Is that the intended outcome?

I struggled to deal with this when creating these variants. I've since realised perhaps using if BOARD_TARGET =  would have been a better way of dealing with some of the variations. Perhaps that would work here. It would be great to be using common definitions.

jfischer-no replied:
fixed

jacobw replied:
Since you have only selected this for the specific boards in Kconfig you don’t need the if here do you?

jfischer-no replied:
Correct, it can now be removed. Thanks.

jacobw commented on boards/seeed/xiao_ble/xiao_ble_nrf52840_sense_defconfig:
```diff
@@ -18,9 +18,6 @@ CONFIG_CONSOLE=y
 # Logger cannot use itself to log
 CONFIG_USB_CDC_ACM_LOG_LEVEL_OFF=y
```
Should this be removed?

jfischer-no replied:
Thanks, removed.

jfischer-no committed:
boards: remove ECM/EEM from default configuration

The board hardware has no network interfaces, although the ECM/EEM class
implementation can provide Ethernet-like functionality and export it to
the host, this is no reason to default to a specific USB class
implementation.We do not make this kind of configuration for other
boards that have a USB device controller. Also, it may not be what a
user expects when using these boards with networking and a USB stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on boards/adafruit/itsybitsy/Kconfig:
```diff
@@ -3,7 +3,7 @@
 
-config BOARD_SERIAL_BACKEND_CDC_ACM
-	bool "USB CDC"
+config BOARD_NEED_SERIAL_BACKEND_CDC_ACM
```
No. This is used in the board context and will stay prefix with `BOARD`. `NEED` is already used in Kconfig files.

tejlmand replied:
> This is used in the board context and will stay prefix with BOARD.

yes in your PR, and my point is that this could potentially be useful in other contexts in the future.
Having `BOARD` as prefix in this case give no added value compared to removing `BOARD`.

> `NEED` is already used in Kconfig files.

I didn't say `NEED` is not used, but I said `REQUIRES` is more used:
> > `REQUIRES` is slightly more used term in Kconfig than `NEED`

Trying to be consistent in Zephyr is important, and in this case trying to align the most used naming is better than on purpose making things different on purpose.

jfischer-no replied:
> yes in your PR, and my point is that this could potentially be useful in other contexts in the future.
> Having `BOARD` as prefix in this case give no added value compared to removing `BOARD`.

No, there actually no future, for all use cases we have cdc-acm-console snippet, it just that hello_world should work "out of the box" without any additional command line arguments.

>Trying to be consistent in Zephyr is important, and in this case trying to align the most used naming is better than on purpose making things different on purpose.

Yes, `git grep "config .*_NEED_"` vs `git grep "config .*_REQUIRED_"`

tejlmand replied:
> No, there actually no future, for all use cases we have cdc-acm-console snippet, 

and you now all the directions the future will take ?
What about the day when it's the SoC or a CPU cluster which will be requiring a specific backend.
cdc acm as serial backend, or a CPU cluster which will requires an IPC serial backend ?

I personally don't know the future, hence I believe it better to avoid the `BOARD` prefix as it doesn't provide any added benefit imo.

> > Trying to be consistent in Zephyr is important, and in this case trying to align the most used naming is better than on purpose making things different on purpose.
> 
> Yes, `git grep "config .*_NEED_"` vs `git grep "config .*_REQUIRED_"`

I said `REQUIRES`, not `REQUIRED`, not to mention that your grep of using `_` in front and after the keyword limits the number of hits. 
The majority use of `NEED` is in the net subsystem, not on the highlevel, where, as example, the `REQUIRES_FULL_LIBC` exists.

nordicjm replied:
We should be using common patterns in tree not having each area use a different terminology, there's different terms used in different things like specifications i.e. "shall have" vs "must have", requires aligns more with that in specifications that I have read, I don't think I've seen any that use need

jfischer-no replied:
> > No, there actually no future, for all use cases we have cdc-acm-console snippet,
> 
> and you now all the directions the future will take ? What about the day when it's the SoC or a CPU cluster which will be requiring a specific backend. cdc acm as serial backend, or a CPU cluster which will requires an IPC serial backend ?
> 
> I personally don't know the future, hence I believe it better to avoid the `BOARD` prefix as it doesn't provide any added benefit imo.

The whole purpose of the first commit is that boards do not want to use the cdc-acm-console snippet because it would require additional build parameters. And CDC ACM UART backend is not selected by this kconfig file, but by the chosen property, Kconfig file is needed because of USB device stack and few other settings for convenience.

> I said `REQUIRES`, not `REQUIRED`, not to mention that your grep of using `_` in front and after the keyword limits the number of hits. The majority use of `NEED` is in the net subsystem, not on the highlevel, where, as example, the `REQUIRES_FULL_LIBC` exists.

okay, changed to BOARD_REQUIRES_SERIAL_BACKEND_CDC_ACM

jfischer-no replied:
If it would be acceptable [Kconfig.cdc_acm_serial.defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/81308/commits/f2b55b00f36484e9452e2f5481346898a51e4563#diff-044af1612b89d9e03c0c285d11e2e5946a1c97bc2bc6a47a0349b3b8e52019c4) to be placed under the `subsys/usb` then I am fine with `REQUIRES_SERIAL_BACKEND_CDC_ACM`, also it could lead to a better integration.

tejlmand replied:
> If it would be acceptable to put [Kconfig.cdc_acm_serial.defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/81308/commits/f2b55b00f36484e9452e2f5481346898a51e4563#diff-044af1612b89d9e03c0c285d11e2e5946a1c97bc2bc6a47a0349b3b8e52019c4) under the `subsys/usb` then I am fine with `REQUIRES_SERIAL_BACKEND_CDC_ACM`, also it might lead to a better integration.

why would that file need to be placed under `subsys/usb` ?
The symbols are defined here: https://github.com/zephyrproject-rtos/zephyr/blob/21bf9253163d565a92d52bfc06d3ad55b4b20bab/boards/Kconfig#L131
https://github.com/zephyrproject-rtos/zephyr/blob/f2b55b00f36484e9452e2f5481346898a51e4563/boards/Kconfig#L137
and those symbols could be moved to `subsys/usb`.

I'm not sure starting to allow `defconfig` files under various subsys / lib / etc. folders is a wise idea.
That will make it even harder to follow from where defconfig files are picked up.

And although defconfig file do adjust many settings defined throughout Zephyr repo, then I also understand the wish to have the definition close to the defconfigs impacted by this setting.
So I also see reasons to keep the symbol definition at current location.

jfischer-no replied:
> > If it would be acceptable to put [Kconfig.cdc_acm_serial.defconfig](https://github.com/zephyrproject-rtos/zephyr/pull/81308/commits/f2b55b00f36484e9452e2f5481346898a51e4563#diff-044af1612b89d9e03c0c285d11e2e5946a1c97bc2bc6a47a0349b3b8e52019c4) under the `subsys/usb` then I am fine with `REQUIRES_SERIAL_BACKEND_CDC_ACM`, also it might lead to a better integration.
> 
> why would that file need to be placed under `subsys/usb` ? The symbols are defined here:

The symbols could be placed there as well. The thing is, snippets cannot contain code, and snippets do not support kconfig files. Boards like the nrf52840 dongle should not use snippets for the CDC ACM serial backend, but have it by default. This means that to switch to the new device support, I need to [add application code](https://github.com/zephyrproject-rtos/zephyr/pull/80985/commits/acbb65d03d9f9c4f0a889f98687c452e6f4bffee) for CDC ACM serial support to subsys/usb. Then I can also put those kconfig options and defaults there.

> I'm not sure starting to allow `defconfig` files under various subsys / lib / etc. folders is a wise idea. That will make it even harder to follow from where defconfig files are picked up.

But this has been done for a long time.

tejlmand replied:
> > I'm not sure starting to allow defconfig files under various subsys / lib / etc. folders is a wise idea. That will make it even harder to follow from where defconfig files are picked up.

> But this has been done for a long time.

What do you mean has been done for a long time ?
Today we only allow defconfig files under `boards` and `socs`, not under lib / subsys.

henrikbrixandersen commented on boards/Kconfig:
```diff
@@ -128,6 +128,11 @@ config QEMU_EXTRA_FLAGS
 	  to setup devices, for example to allocate interface for Zephyr
 	  GDBstub over serial with `-serial tcp:127.0.0.1:5678,server`
 
+config BOARD_SERIAL_BACKEND_CDC_ACM
+	bool "Board uses USB CDC ACM UART as serial backend"
```
Segger RTT could be added in the same way later on.

jfischer-no replied:
While using CDC ACM as the default backend for the boards without debug adapter, but with USB device controller and **bootloader** makes sense, what would be the case for "Segger RTT"? Is Segger RTT a subsystem that needs some configuration and how would that look like?

henrikbrixandersen replied:
We currently have this configuration available as a snippet (https://github.com/zephyrproject-rtos/zephyr/tree/main/snippets/rtt-console) which works fine for me. It was merely an example of another "serial" backend, that might be added like this in the future.

jfischer-no replied:
Yes, and we know that there is also a cdc-acm-console snippet. But the requirements for this work are to make samples like "hello_world" work "out of the box" on boards with USB device controller and pre-flashed bootloader like nrf52840dongle. Are there similar use cases for boards with a Segger debug adapter? And if so, what would it look like? Is it worth adding support for this in boards/common and duplicate rtt-console snippet?

henrikbrixandersen replied:
I don't know if this need exists for in-tree boards, but for out-of-tree boards, it would certainly be a nice addition.

tejlmand replied:
> it would be against "General recommendations"

why would it be against "General recommendations" to have different serial backends ?

Enabling a different serial backend per default in upstream could be going against "General recommendations", but not having a choice which provides an option to select another backend.
A choice would furthermore ensure that only a single backend can be chosen at a time, like a user doing `-DCONFIG_SERIAL_BACKEND_FOO=y` will ensure that the CDC_ACM serial backend is not being selected.
For example when select RTT console.

jfischer-no replied:
"Unless explicitly recommended otherwise by this section, leave peripherals and their drivers disabled by default." and "If available, configure and enable a serial output for the console..." IMO the wording only considers a real UART controller as a serial output.

>like a user doing -DCONFIG_SERIAL_BACKEND_FOO=y will ensure that the CDC_ACM serial backend is not being selected.
For example when select RTT console.

User doing -DCONFIG_SERIAL_BACKEND_FOO=y is completely wrong here, user capable to use build arguments should use `-DSNIPPET="rtt_console"` or `-DSNIPPET="cdc-acm-console"`. Backend configuration at the board level must be set to a single specific backend, if at all. No choices. If we want choices then snippets should be used and I can close this PR.

nordicjm replied:
I consider USB CDC to be a serial output device, and from my prior job as a vendor that had things like USB dongles added to zephyr, the company opinion was that they too are serial output devices and should be enabled by default

tejlmand replied:
> is completely wrong here, user capable to use build arguments should use -DSNIPPET="rtt_console" or -DSNIPPET="cdc-acm-console".

and what does `-DCONFIG_SERIAL_BACKEND_FOO` vs `-DSNIPPET=<foo>` selecting the backend have to do with the selecting being a choice or not ?

Even if this is configured through a snippet the Kconfig symbol should still be a choice.

jfischer-no replied:
> and what does `-DCONFIG_SERIAL_BACKEND_FOO` vs `-DSNIPPET=<foo>` selecting the backend have to do with the selecting being a choice or not ?

`-DSNIPPET="cdc-acm-console"` selects logging serial backend, `-DCONFIG_BOARD_SERIAL_BACKEND_CDC_ACM=y` does not, it just enables USB device stack and applies few optional default settings. For this PR, **the board has already chosen the backend** using chosen property `zephyr,console` or `zephyr,shell-uart`, `boards/common/usb/Kconfig.cdc_acm_serial.defconfig` are dependencies for that choice.

tejlmand replied:
if this setting is not intended for the user to select, then the prompt should be removed.

carlescufi reviewed (commented):
+1 for the approach

bryceschober commented on boards/st/steval_stwinbx1/doc/index.rst:
```diff
@@ -230,23 +230,7 @@ Console
-                compatible = "zephyr,cdc-acm-uart";
-        };
-     };
+- through common CDC ACM UART backend configuraiton for all boards
```
This typo caught my eye, and has at least one other instance in this PR.
```suggestion
- through common CDC ACM UART backend configuration for all boards
```

tejlmand reviewed (dismissed):
lgtm.

Still some minor details, but those are very much depending on the perspective from where the solution is seen, so not something which should block this PR.

nordicjm commented on boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:
```diff
@@ -28,5 +28,6 @@ CONFIG_SERIAL=y
 
-# Enable USB Stack (needed for the console to work)
-CONFIG_USB_DEVICE_STACK=y
+# Enable regulator (needed to enable eth)
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_FIXED=y
```
is this change correct?

nordicjm commented on boards/makerdiary/nrf52840_mdk_usb_dongle/Kconfig:
```diff
@@ -4,10 +4,13 @@
 # SPDX-License-Identifier: Apache-2.0
 
+if BOARD_NRF52840_MDK_USB_DONGLE
+
 config BOARD_HAS_NRF5_BOOTLOADER
 	bool "Board has nRF5 bootloader"
 	default y
-	depends on BOARD_NRF52840_MDK_USB_DONGLE
 	help
 	  If selected, applications are linked so that they can be loaded by Nordic
 	  nRF5 bootloader.
+
+endif # BOARD_NRF52840_MDK_USB_DONGLE
```
this change is pointless

nordicjm commented on boards/nordic/thingy53/thingy53_nrf5340_cpuapp_ns.dts:
```diff
@@ -7,6 +7,8 @@
 #include <nordic/nrf5340_cpuappns_qkaa.dtsi>
 #include "thingy53_nrf5340_common.dtsi"
+#include <../boards/common/usb/cdc_acm_serial.dtsi>
+
```
double newline
